infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -0,0 +1,514 @@
+        this.pluginFilePath = serverContext.getPluginConfigFilePath();
+        this.discoveryCallback = discoveryCallback;
+    }
+
+    public void run() {
+        try {
+            startDiscovery();
+
+            while (shouldRun) {
+                try {
+                    DiscoveryServiceEvent event = eventQueue.take();
+                    processEvent(event);
+                } catch (Exception e) {
+                    log.error("Caught an exception. Stop discovery service.", e);
+                    shouldRun = false;
+                    stopLogReplication();
+                    if (e instanceof InterruptedException) {
+                        Thread.interrupted();
+                    }
+                }
+            }
+        } catch (LogReplicationDiscoveryServiceException e) {
+            log.error("Exceptionally terminate Log Replication Discovery Service", e);
+            discoveryCallback.completeExceptionally(e);
+        } catch (Exception e) {
+            log.error("Unhandled exception caught during log replication service discovery", e);
+        } finally {
+            if (runtime != null) {
+                runtime.shutdown();
+            }
+        }
+    }
+
+    /**
+     * On first access start topology discovery.
+     *
+     * On discovery, process the topology information and fetch log replication configuration
+     * (streams to replicate) required by an active and standby site before starting
+     * log replication.
+     */
+    private void startDiscovery() throws LogReplicationDiscoveryServiceException {
+
+        try {
+            log.info("Connect to Cluster Manager adapter.");
+
+            this.clusterManagerAdapter.connect(this);
+
+            log.info("Fetch topology from Cluster Manager...");
+
+            topologyDescriptor = new TopologyDescriptor(clusterManagerAdapter.fetchTopology());
+
+            // Health check - confirm this node belongs to a cluster in the topology
+            if (clusterPresentInTopology(topologyDescriptor)) {
+
+                log.info("Node[{}] belongs to cluster, descriptor={}", localEndpoint,
+                        localClusterDescriptor);
+
+                LogReplicationContext context = buildLogReplicationContext();
+
+                // Unblock server initialization retrieving context: topology + configuration
+                discoveryCallback.complete(context);
+
+                registerToLogReplicationLock();
+            } else {
+                // If a cluster descriptor is not found, this node does not belong to any topology... raise an exception
+                String message = String.format("Node[%s] does not belong to any Cluster provided by the discovery service, topology=%s",
+                        localEndpoint, topologyDescriptor);
+                log.warn(message);
+                throw new LogReplicationDiscoveryServiceException(message);
```

Comment:
This is a valid point, similar to how we retry registration to lock until successful. We can think of permitting subscription to Cluster Manager until we appear as part of the cluster and it might make sense in the event of a topology being formed. I've added this and your other comment of the lock for discussion and we can push them in new PRs.